File: day1/part1/expenseReport.py
from timetaken import averagetime, timetaken

f = open("input", "r")
expenses = f.readlines()
expenses = list(map(lambda e: int(e), expenses))
n_expenses = len(expenses)

def bruteforce():
  for i in range(n_expenses):
    for j in range(n_expenses):
      if (i != j and expenses[i] + expenses[j] == 2020):
        return expenses[i] * expenses[j]

def sortedfirst():
  expenses.sort()
  for i in range(n_expenses):
    for j in reversed(range(n_expenses)):
      if (expenses[i] + expenses[j] < 2020):
        continue
      if (i != j and expenses[i] + expenses[j] == 2020):
        return expenses[i] * expenses[j]

print("Brute Force")
timetaken(bruteforce)
averagetime(bruteforce)

print("Sorted")
timetaken(sortedfirst)
averagetime(sortedfirst)

print(sortedfirst())

File: day1/part1/expenseReport.js
const fs = require('fs');

let expenseData;
const expenses = [];
try {
  expenseData = fs.readFileSync('./input', 'utf8');
  curExpense = '';
  for(let i = 0; i < expenseData.length; i += 1) {
    if(expenseData[i] === '\n') {
      expenses.push(parseInt(curExpense, 10));
      curExpense = '';
    } else {
      curExpense += expenseData[i];
    }
  }
} catch (err) {
  console.error(err);
}

expenses.sort((a, b) => a - b);
nExpenses = expenses.length;

const expenseProduct = () => {
  for(let i = 0; i < nExpenses; i += 1) {
    for(let j = nExpenses - 1; j > 0; j -= 1) {
      if (expenses[i] + expenses[j] === 2020) {
        console.log(expenses[i] * expenses[j]);
        return;
      }
      if (expenses[i] + expenses[j] < 2020) {
        break;
      }
    }
  }
}

expenseProduct();
